pkg/gardenlet/controller/shoot/seed_registration_control.go

```diff
@@ -113,884 +68,298 @@ func (c *Controller) seedRegistrationUpdate(oldObj, newObj interface{}) {
-			Domain:     ingressDomain,
-			Controller: *shootedSeedConfig.IngressController,
-		}
-	} else {
-		dns.IngressDomain = &ingressDomain
-	}
-
-	var loadBalancerServices *gardencorev1beta1.SeedSettingLoadBalancerServices
-	if shootedSeedConfig.LoadBalancerServicesAnnotations != nil {
-		loadBalancerServices = &gardencorev1beta1.SeedSettingLoadBalancerServices{
-			Annotations: shootedSeedConfig.LoadBalancerServicesAnnotations,
-		}
-	}
-
-	return &gardencorev1beta1.SeedSpec{
-		Provider: gardencorev1beta1.SeedProvider{
-			Type:           shoot.Spec.Provider.Type,
-			Region:         shoot.Spec.Region,
-			ProviderConfig: shootedSeedConfig.SeedProviderConfig,
-		},
-		DNS:       dns,
-		Ingress:   ingress,
-		SecretRef: secretRef,
-		Networks: gardencorev1beta1.SeedNetworks{
-			Pods:          *shoot.Spec.Networking.Pods,
-			Services:      *shoot.Spec.Networking.Services,
-			Nodes:         shoot.Spec.Networking.Nodes,
-			BlockCIDRs:    shootedSeedConfig.BlockCIDRs,
-			ShootDefaults: shootedSeedConfig.ShootDefaults,
-		},
-		Settings: &gardencorev1beta1.SeedSettings{
-			ExcessCapacityReservation: &gardencorev1beta1.SeedSettingExcessCapacityReservation{
-				Enabled: shootedSeedConfig.DisableCapacityReservation == nil || !*shootedSeedConfig.DisableCapacityReservation,
-			},
-			Scheduling: &gardencorev1beta1.SeedSettingScheduling{
-				Visible: shootedSeedConfig.Visible == nil || *shootedSeedConfig.Visible,
-			},
-			ShootDNS: &gardencorev1beta1.SeedSettingShootDNS{
-				Enabled: shootedSeedConfig.DisableDNS == nil || !*shootedSeedConfig.DisableDNS,
-			},
-			LoadBalancerServices: loadBalancerServices,
-			VerticalPodAutoscaler: &gardencorev1beta1.SeedSettingVerticalPodAutoscaler{
-				Enabled: vpaEnabled,
-			},
+func deleteManagedSeed(ctx context.Context, c client.Client, shoot *gardencorev1beta1.Shoot) error {
+	// Delete managed seed
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      shoot.Name,
+			Namespace: shoot.Namespace,
 		},
-		Taints: taints,
-		Backup: backupProfile,
-		Volume: volume,
-	}, nil
-}
-
-func activateSeedIngressController(ctx context.Context, seedClient client.Client, shoot *gardencorev1beta1.Shoot, ingressController *gardencorev1beta1.IngressController) (bool, error) {
-	if gardencorev1beta1helper.NginxIngressEnabled(shoot.Spec.Addons) {
-		return false, nil
-	}
-
-	if seedIngressEnabled := ingressController != nil; !seedIngressEnabled {
-		return false, nil
 	}
-
-	// If migrating from Shoot Nginx Addon to Seed Ingress Controller wait until DNSEntry pointing to ShootNginxAddon service got deleted before activating Seed Ingress Controller.
-	if err := seedClient.Get(ctx, kutil.Key(shoot.Status.TechnicalID, common.ShootDNSIngressName), &dnsv1alpha1.DNSEntry{}); err != nil {
-		if apierrors.IsNotFound(err) {
-			return true, nil
-		}
-		return false, err
+	if err := c.Delete(ctx, managedSeed); client.IgnoreNotFound(err) != nil {
```

Comment:
I don't think so. The shoots must be removed first, before the seed is actually deleted, this is ensured by the seed reconciliation logic. The managed seed itself won't be deleted unless the seed is already deleted, and will report associated objects, including shoots, see https://github.com/stoyanr/gardener/blob/a244064898b4436678000f15b7d090937e7a0ddf/pkg/gardenlet/controller/managedseed/managedseed_actuator.go#L210.

It's been a while since I last tested this though, so let me double check it still works as expected, you could also try it out.

Reply:
Right, makes sense. Just wanted to double-check.
I will give this scenario a try when playing around with it locally and report if I see any issues with the current implementation.

Reply:
I tested this, it works as described. The managed seed is marked for deletion, but nothing really happens if there is a shoots still scheduled on the seed - the seed deletion is rejected, and the deletion of the managed seed fails. If the shoot is deleted, the seed deletion proceeds and the managed seed is eventually also deleted.

Sounds like good enough for me, if you think anything should be changed please let me know.